src/Interpreters/DatabaseCatalog.h

```diff
@@ -235,6 +235,29 @@ class DatabaseCatalog : boost::noncopyable, WithMutableContext
 
     void checkTableCanBeRemovedOrRenamed(const StorageID & table_id, bool check_referential_dependencies, bool check_loading_dependencies, bool is_drop_database = false) const;
 
+
+    struct TableMarkedAsDropped
+    {
+        StorageID table_id = StorageID::createEmpty();
+        StoragePtr table;
+        String metadata_path;
+        time_t drop_time{};
+    };
+    using TablesMarkedAsDropped = std::list<TableMarkedAsDropped>;
+
+    struct TableMarkedDroppedForSystemTable
+    {
+        UInt32 index;
+        String database;
+        String table;
+        UUID uuid;
+        String engine;
+        String metadata_dropped_path;
+        DateTime64 table_dropped_time;
+    };
```

Comment:
Why not use TableMarkedAsDropped? (new structure does not fix the issue with nullptr)